### PR TITLE
feat(web): start of input boundary generalization 🐵

### PR DESCRIPTION
**Pull request by jahorton**
The `VisualKeyboard` class has some... fun logic for determining when to cancel an ongoing touch or mouse interaction.  Input may continue slightly out of bounds - but not too far.  Input that gets too close to a screen edge is also interpreted as cancellation (since you can't detect past physical device boundaries) - unless the interaction started very near to those boundaries.

After some thought, analysis, and effort... I've come up with a scheme that will allow this to be specified relatively cleanly - both within this module and within KeymanWeb once integrated.  This scheme will _also_ allow for other boundary specifications - like those that will be needed for a gesture testing host/demo page.

@keymanapp-test-bot skip

### CODE REVIEW

File: common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts [hunk 1]
```diff
@@ -18,6 +25,45 @@ namespace com.keyman.osk {
...
+     * start within this pixel range from a safe bound will disable that bound for the duration
+     * of its corresponding input sequence.  May be a number or an array of 1, 2, or 4 numbers,
+     * as with CSS styling.
+     *
+     * If not specified, this will default to a padding of 3px inside the standard safeBounds.
+     */
+    readonly safeBoundPadding?: number | number[];
+  }
+
+  export interface FinalizedGestureRecognizerConfiguration extends Nonoptional<GestureRecognizerConfiguration> {
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts:60]: Is there a real need for a separate interface? It seems like it adds a lot of boilerplate.

File: common/web/gesture-recognizer/src/mouseEventEngine.ts [hunk 2]
```diff
@@ -79,7 +78,13 @@ namespace com.keyman.osk {
       }
 
       this.preventPropagation(event);
-      this.onInputStart(InputEventCoordinate.fromEvent(event));
+      const coord = InputEventCoordinate.fromEvent(event);
+
+      if(!ZoneBoundaryChecker.inputStartOutOfBoundsCheck(coord, this.config)) {
+        this.disabledSafeBounds = ZoneBoundaryChecker.inputStartSafeBoundProximityCheck(coord, this.config);
+      }
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/mouseEventEngine.ts:85]: What is this doing? It's not obvious why you are doing this -- please add a code comment!

File: common/web/gesture-recognizer/src/paddedZoneSource.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            this.edgePadding = [edgePadding[0],
+                                edgePadding[1],
+                                edgePadding[0] + edgePadding[2],
+                                2*edgePadding[1]];
+          case 4:
+            // top, right, bottom, left
+            this.edgePadding = [edgePadding[0],
+                                edgePadding[3], // we want the `left` entry internally, not the `right`.
+                                edgePadding[0] + edgePadding[2],
+                                edgePadding[1] + edgePadding[3]];
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/paddedZoneSource.ts:49]: So the input of this function is top, right, bottom, left, but the internal member is top, left, height, width? Seems likely to cause us to trip up in the future? (But if we change the internal member to an object as I suggest, this problem goes away.)

File: common/web/gesture-recognizer/src/paddedZoneSource.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  export class PaddedZoneSource implements RecognitionZoneSource {
+    private readonly root: RecognitionZoneSource;
+
+    /**
+     * [0]: y (top)
+     * [1]: x (left)
+     * [2]: height (top+bottom)
+     * [3]: width (left+right)
+     */
+    private edgePadding: number[];
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/paddedZoneSource.ts:12]: I think this should be an object, not an array:

```
private edgePadding: { x: number, y: number, width: number, height: number };
```

File: common/web/gesture-recognizer/src/zoneBoundaryChecker.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+     * boundaries) the input coordinate lies outside of.
+     *
+     * @param coord         An input coordinate
+     * @param zone          An object defining a 'recognition zone' of the gesture engine.
+     * @param ignoreBitmask A bitmask indicating select boundaries to ignore for the check.
+     */
+    static getCoordZoneBitmask(coord: InputEventCoordinate, zone: RecognitionZoneSource, ignoreBitmask?: number): number {
+      if(!ignoreBitmask) {
+        ignoreBitmask = 0;
+      }
```
Comment (mcdurdin) [inline common/web/gesture-recognizer/src/zoneBoundaryChecker.ts:23]: Given `ignoreBitmask` is only used in `inputMoveCancellationCheck()`, perhaps move the logic for it there and simplify the purpose of this function?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
There seems to be some code repetition which should be addressed. A number of other nits.

File: common/web/gesture-recognizer/src/mouseEventEngine.ts (hunk 2 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/mouseEventEngine.ts:85]: This comment is a secondary target of this comment: https://github.com/keymanapp/keyman/pull/6843#discussion_r919571361

This addresses probably the most niche aspect of the OSK's touchpoint management logic, so I'm not exactly surprised by the request.  Re #6842's visual:  "If the touchpoint is within the larger set of dashed lines but not the smaller set, remember which side we started on."  And _that_ has a second half.  Not exactly the simplest thing to document using only words; this is part of why I spent as much time as I did on the host page's visual design.

File: common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts (hunk 1 above)
Comment (mcdurdin) [reply common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts:57]: It's just a little too OO for my comfort -- why not just make `paddedSafeBounds` optional on the input object as you say with corresponding comments? It's all internal anyway so it's not like we are making API contracts.

**Review by mcdurdin — APPROVED** [review]

File: common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts (hunk 1 above)
Comment (jahorton) [reply common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts]: I've made an adjustment in this regard; are the recent changes enough to address this concern?

Comment (mcdurdin) [reply common/web/gesture-recognizer/src/gestureRecognizerConfiguration.ts]: Yes thanks

**Pull request merged**